### DISCUSSION TITLE
How do i use dynamic replacements with translatables?

### DISCUSSION THREAD

**Discussion by L3v3lup** [KyoriPowered/adventure#944] [General]
Hello,

i need some help with translatables and dynamic replacements. I have gone through the docs multiple times but i can't get it to work.
The "best" version i have is:
```Java
Component component = Component.translatable("item.minecraft." + itemStack.getType().getKey().getKey());
final Component p2 = MiniMessage.miniMessage().deserialize("<lang:> <translatable>", Placeholder.component("translatable", component));
Bukkit.broadcast(p2);
```
The output this gives me is: `item.minecraft.ancient_debris`.
I have tried dozens of variants but i can't get it to work. What am i doing wrong? Please help me out with that :)

**Comment by Machine-Maker** [comment]
I'm assuming you are using Paper. You seem to be making it more complex than needed.

```java
@EventHandler
public void onEvent(ChatEvent event) {
    final TranslatableComponent component = Component.translatable(new ItemStack(Material.ANCIENT_DEBRIS));
    final Component mini = MiniMessage.miniMessage().deserialize("<translatable>", Placeholder.component("translatable", component));
    event.getPlayer().sendMessage(mini);
}
```

This seems to do what you want, and works fine for me.

**Comment by L3v3lup** [comment]
Thanks for your message.
I am using Paper (Purpur 1.20.1 to be exact), i forgot to mention that.
That is working fine, thank you so much!

I just figured out that it is also important, to use the correct "sendMessage()" method from Kyori (not Bukkit).

**Comment by L3v3lup** [comment]
I do have another question.
How do i include a hover event to that translated component? I would like to also display the item in a Hover-Event.
To be exact; I am displaying a server message / broadcast to all players, when they find one of the new smithing templates. Since they have the same name, i would like to add a hover-event to display the whole item.

I dont have an idea where/how to inculde that. Do i need to make another Placeholder? Or is it some kind of component? 
Sorry, i still dont understand that API yet.

**Reply by Machine-Maker** [reply]
Just add a hover event to the component created by `Component.translatable` and use that in the placeholder like you already are. 
```java
Component someComponent;
someComponent.hoverEvent(new ItemStack(Material.ANCIENT_DEBRIS));
```

**Reply by L3v3lup** [reply]
Oh that is much easier than i thought. Thank you so much :)
